### PR TITLE
Fix possible double-free in pkcs7 add_attribute function

### PR DESCRIPTION
**Pull request by bernd-edlinger**
The problem is the ownership of the input parameter value is transfered to the X509_ATTRIBUTE object attr, as soon as X509_ATTRIBUTE_create succeeds, but when an error happens after that point there is no way to get get the ownership back to the caller, which is necessary to fullfill the API contract.

Fixed that by moving the call to X509_ATTRIBUTE_create to the end of the function, and make sure that no errors are possible after that point.

### CODE REVIEW

**Comment by bernd-edlinger** [comment]
This PR will need to do some more re-work...

**Comment by bernd-edlinger** [comment]
Error handling of X509_ATTRIBUTE_create fixed.

**Review by t8m — APPROVED** [review]

File: crypto/pkcs7/pk7_doit.c
Comment (tmshort) [inline crypto/pkcs7/pk7_doit.c:1260]: The prior code checked the return value, this doesn't, which _could_ be a leakof `attr`, although I suspect that it would never actually fail (since the stack doesn't need to grow)?

Comment (bernd-edlinger) [reply crypto/pkcs7/pk7_doit.c:1260]: Yeah, the idea is that it cannot fail because this element was
either pre-existing, or successfully created by `sk_X509_ATTRIBUTE_push(*sk, NULL)`
above.  Any error after the point-of-no-return would be fatal in any case.

**Comment by t8m** [comment]
ping @openssl/committers for second review

**Review by nhorman — APPROVED** [review]

**Comment by t8m** [comment]
Merged to all the active branches. Thank you for your contribution.

**Pull request closed without merging**
